Nodejs/Product/Npm/PackageJsonFactory.cs

```diff
@@ -1,14 +1,109 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
+using System.Globalization;
+using System.IO;
+using System.Threading;
 using Microsoft.NodejsTools.Npm.SPI;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.NodejsTools.Npm
 {
-    public class PackageJsonFactory
+    public static class PackageJsonFactory
     {
-        public static IPackageJson Create(IPackageJsonSource source)
+        public static IPackageJson Create(string fullPathToFile)
         {
-            return source.Package == null ? null : new PackageJson(source.Package);
+            if (string.IsNullOrEmpty(fullPathToFile) && !fullPathToFile.EndsWith("package.json"))
+            {
+                throw new ArgumentException("Expected full path to 'package.json' file.", nameof(fullPathToFile));
+            }
+
+            if (File.Exists(fullPathToFile))
+            {
+                var retryInterval = 500;
+                var attempts = 5;
+
+                // populate _source with retries for recoverable errors.
+                while (--attempts >= 0)
+                {
+                    try
+                    {
+                        using (var fin = new FileStream(fullPathToFile, FileMode.Open, FileAccess.Read, FileShare.Read))
+                        using (var reader = new StreamReader(fin))
+                        {
+                            return Create(reader);
+                        }
+                    }
+                    catch (PackageJsonException pje)
+                    {
+                        WrapExceptionAndRethrow(pje);
+                    }
+                    catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
+                    {
+                        if (attempts <= 0) { throw; }
+                    }
+
+                    Thread.Sleep(retryInterval);
+                    retryInterval *= 2; // exponential backoff
+                }
+            }
+
+            return null;
+
+            void WrapExceptionAndRethrow(Exception ex)
+            {
+                throw new PackageJsonException(
+                    string.Format(CultureInfo.CurrentCulture, @"Error reading package.json at '{0}': {1}", fullPathToFile, ex.Message),
```

Comment:
Probably redundant to say "package.json" when it's going to appear at the end of the path anyway.